core/src/main/java/hudson/PluginManager.java

```diff
@@ -1341,41 +1343,55 @@ public static boolean isNonMetaLabel(String label) {
     
     @Restricted(NoExternalUse.class)
     public HttpResponse doPluginsSearch(@QueryParameter String query, @QueryParameter Integer limit) {
+        final String _query = query.toLowerCase(Locale.ROOT);
+        final boolean isWarningsQuery = _query.contains("warning:");
         List<JSONObject> plugins = new ArrayList<>();
         for (UpdateSite site : Jenkins.get().getUpdateCenter().getSiteList()) {
             List<JSONObject> sitePlugins = site.getAvailables().stream()
                 .filter(plugin -> {
-                    if (StringUtils.isBlank(query)) {
+                    // preserve an existing feature for the "warning:" query
+                    if (isWarningsQuery && plugin.hasWarnings()) {
                         return true;
                     }
-                    return StringUtils.containsIgnoreCase(plugin.name, query) ||
-                        StringUtils.containsIgnoreCase(plugin.title, query) ||
-                        StringUtils.containsIgnoreCase(plugin.excerpt, query) ||
-                        plugin.hasCategory(query) ||
-                        plugin.getCategoriesStream()
-                            .map(UpdateCenter::getCategoryDisplayName)
-                            .anyMatch(category -> StringUtils.containsIgnoreCase(category, query)) ||
-                        plugin.hasWarnings() && query.equalsIgnoreCase("warning:");
-                })
-                .limit(Math.max(limit - plugins.size(), 1))
-                .sorted((o1, o2) -> {
-                    String o1DisplayName = o1.getDisplayName();
-                    if (o1.name.equalsIgnoreCase(query) ||
-                        o1DisplayName.equalsIgnoreCase(query)) {
-                        return -1;
-                    }
-                    String o2DisplayName = o2.getDisplayName();
-                    if (o2.name.equalsIgnoreCase(query) || o2DisplayName.equalsIgnoreCase(query)) {
-                        return 1;
+                    // if the query is a category, return plugins in that category
+                    if (plugin.hasCategory(_query)) {
```

Comment:
💡 We should provide a user feedback that they chose a category (with information that you can do it), otherwise such feature will not be used voluntarily, leading to confusion without added value until you have understood what is going on. (= not really good UX)

Reply:
New implicit requirement is that categories must be all lowercase. I don't think that was the case before.

Reply:
@daniel-beck I modified #hasCategory in UpdateSite.java so that the comparison is done in lowercase on both sides. I'm not sure if there are side effects, as I haven't run a full test suite yet.

Reply:
categories are generally used either by clicking on them from other plugins or as links from other places in the UI,

e.g. no cloud plugins installed, click here to find a cloud plugin to install